programs/whirlpool/Cargo.toml

```diff
@@ -1,37 +1,36 @@
 [package]
 name = "whirlpool"
-version = "0.3.0"
-description = "Created with Anchor"
-edition = "2018"
+version = "0.3.1"
+edition = "2021"
 
 [lib]
 crate-type = ["cdylib", "lib"]
-name = "whirlpool"
 
 [features]
 no-entrypoint = []
 no-idl = []
 cpi = ["no-entrypoint"]
 default = []
+idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
 
 [dependencies]
-anchor-lang = "0.29"
-anchor-spl = {version = "0.29", features = ["metadata", "memo"]}
-spl-token = {version = "4", features = ["no-entrypoint"]}
-spl-transfer-hook-interface = "0.5.0"
-solana-program = "1.17"
-thiserror = "1.0"
-uint = {version = "0.9.1", default-features = false}
-borsh09 = {package = "borsh", version = "0.9.1"}
-arrayref = "0.3.7"
-bytemuck = "1.14.3"
+anchor-lang = { version = "=0.30.1" }
+anchor-spl = { version = "=0.30.1", features = ["metadata", "memo"] }
+spl-token = { version = "=4.0.3", features = ["no-entrypoint"] }
+spl-transfer-hook-interface = { version = "=0.6.5" }
+solana-program = { version = "=1.18.21" }
+thiserror = { version = "=1.0.63" }
+uint = { version = "=0.9.5", default-features = false }
+arrayref = { version = "=0.3.8" }
+borsh09 = { package = "borsh", version = "=0.9.3" }
 solana-security-txt = { version = "=1.1.1" }
+bytemuck = { version = "=1.16.3", features = ["derive", "min_const_generics"] }
 
 [dev-dependencies]
-proptest = "1.0"
-serde = "1.0.117"
-serde_json = "1.0.59"
+proptest = "=1.5.0"
+serde = "=1.0.205"
+serde_json = "=1.0.122"
```

Comment:
This was done because in #172 I added more rust packages in the workspace. These might have different versions of the same libraries. I need to pin them here to the exact versions otherwise it might take a version that is not compatible (but would still satisfy the version constraint here.

The dev dependencies don't matter that much but I figured might as well pin them too. IMO the more of the 'core' dependencies we can pin the less chance there is that in the future it takes a different version which breaks things. Dependabot should ping us when new versions become available which IMO is more transparent than cargo doing those things under the hood.

Reply:
hmm, I think the Rust SDK gets requests from developers who want to use with Anchor 0.29, Solana 1.17 or something more up-to-date.
That's what the crate for CPI was. Once anchor released v0.30.0, some devs started asking to update whirlpools crate.
In light of such requests, I think the contract and SDK versions should be intentionally separated.
I think the workspace that ties those two together can be problematic, what do you think?

Reply:
Yep agree. If you look at #172 you can see the versions for the rust sdk are more permissive. This is only to prevent any local version mismatches. Any downstream consumer would be able to specify versions they would like to use as long as it complies with the constraints in the rust-sdk cargo file. At least that is how I intended it to be